derive/src/lib.rs

```diff
@@ -0,0 +1,48 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+use proc_macro2::TokenStream as TokenStream2;
+use quote::quote;
+use syn::{Data, DeriveInput, Fields};
+
+#[proc_macro_derive(FromArgs)]
+pub fn derive_from_args(input: TokenStream) -> TokenStream {
+    let ast: DeriveInput = syn::parse(input).unwrap();
+
+    let gen = impl_from_args(&ast);
+    gen.to_string().parse().unwrap()
+}
+
+fn impl_from_args(input: &DeriveInput) -> TokenStream2 {
+    // FIXME: This references types using `crate` instead of `rustpython_vm`
+    //        so that it can be used in the latter. How can we support both?
+    let fields = match input.data {
+        Data::Struct(ref data) => {
+            match data.fields {
+                Fields::Named(ref fields) => fields.named.iter().map(|field| {
+                    let name = &field.ident;
+                    quote! {
+                        #name: crate::pyobject::TryFromObject::try_from_object(
+                            vm,
+                            args.take_keyword(stringify!(#name)).unwrap_or_else(|| vm.ctx.none())
+                        )?,
+                    }
+                }),
+                Fields::Unnamed(_) | Fields::Unit => unimplemented!(), // TODO: better error message
+            }
+        }
+        Data::Enum(_) | Data::Union(_) => unimplemented!(), // TODO: better error message
+    };
+
+    let name = &input.ident;
+    quote! {
+        impl crate::function::FromArgs for #name {
```

Comment:
As this crate clearly depends upon rustpython_vm, I would not make a new crate. This makes the code less easy to understand. Why would you create a new crate?

Reply:
That's required to make a procedural macro. If you look in it's Cargo.toml, you'll see `proc-macro = true`, which means that you can export proc macros from it, while you can't do that from a normal crate.

Reply:
Also note it _doesn't_ depend on `rustpython_vm` (as I originally thought it would) - This is inside a `quote!` macro so it's just generating it on the syntax tree level.

Reply:
Can't we set `proc-macro = true` in the rustpython_vm crate?

Reply:
No, rust is pretty strict about the separation of proc macros:

```
error: `proc-macro` crate types cannot export any items other than functions tagged with `#[proc_macro_derive]` currently
  --> vm/src/lib.rs:58:1
   |
58 | pub use self::vm::VirtualMachine;
```

Reply:
Is there another way in which we can prevent the new crate? Why do we require the procedural macro?

Reply:
No, I think this is necessary to have fully accurate argument validation for all cases. Right now, neither the `arg_check!` macro or the new style property handle some cases, such as arguments that can be provided either as positional or keyword.

With the macro, we can just do the following:

```rust
#[derive(FromArgs)]
struct CustomArgs {
    #[pyarg(kind = "positional_or_keyword")]
    arg1: Option<PyObjectRef>,
    arg2: ...
}
```

The macro will create a `FromArgs` implementation for this struct that handles all of this complexity, efficiently, and prevents it from being duplicated for every function that accepts arguments this way.

(It's also the idiomatic rust way to handle cases like this.)

Reply:
The nice thing is that if we want to make another procedural macro, we don't have to make a new crate - all of them can go into `rustpython_derive`.

Reply:
@windelbouwman What are your concerns about adding a new crate?

Reply:
My concerns are that the code becomes less readable. The other concern is the splitting of software into seperate packages in general. Have you heard about leftpad in javascript land? Splitting software into multiple pieces is not always a good idea. That being said, I think for rust python, the software could be split into parser, compiler, vm and bytecode for example. Logical chunks, but I am also in favor of a monolithic rustpython crate which contains all.

That being said, I think we can go forward with this change. I'd rather not, but if there is no other way, we should do it!

Reply:
> My concerns are that the code becomes less readable.

I'll add docs on deriving `FromArgs` to `vm/src/function.rs` that explain how this works, so hopefully that will help mitigate this concern.

> Have you heard about leftpad in javascript land?

Yeah, I remember than incident!

Did you know there's also a lib called `is-even`, with [this implementation](https://github.com/jonschlinkert/is-even/blob/master/index.js#L13)? :rofl:  

> That being said, I think for rust python, the software could be split into parser, compiler, vm and bytecode for example. 

Sounds reasonable!

> but I am also in favor of a monolithic rustpython crate which contains all.

Many projects in rust do both, i.e. have a "monolithic" crate, that actually just re-exports the constituent crates to create a cohesive experience for users. The [new `gloo` project](https://rustwasm.github.io/2019/03/12/lets-build-gloo-together.html) takes this approach for example (see the point about creating an umbrella crate).

> That being said, I think we can go forward with this change. I'd rather not create the extra crate, but if there is no other way, we should do it!

Thanks! I'll do what I can to mitigate your concerns. FWIW, I'm not a fan of rust forcing proc-macros into their own crates either - hopefully they'll find a way to make that unnecessary in the future.